source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc

```diff
@@ -57,39 +62,71 @@ UdpProxyFilter::ActiveSession::ActiveSession(UdpProxyFilter& parent,
   // handle.
 }
 
+UdpProxyFilter::ActiveSession::~ActiveSession() {
+  parent_.config_->stats().downstream_sess_active_.dec();
+}
+
+void UdpProxyFilter::ActiveSession::onIdleTimer() {
+  ENVOY_LOG(debug, "session idle timeout: downstream={} local={}", addresses_.peer_->asStringView(),
+            addresses_.local_->asStringView());
+  parent_.config_->stats().idle_timeout_.inc();
+  parent_.sessions_.erase(addresses_);
+}
+
 void UdpProxyFilter::ActiveSession::onReadReady() {
-  // TODO(mattklein123): Refresh idle timer.
+  idle_timer_->enableTimer(parent_.config_->sessionTimeout());
+
+  // TODO(mattklein123): We should not be passing *addresses_.local_ to this function as we are
+  //                     not trying to populate the local address for received packets.
   uint32_t packets_dropped = 0;
   const Api::IoErrorPtr result = Network::Utility::readPacketsFromSocket(
       *io_handle_, *addresses_.local_, *this, parent_.config_->timeSource(), packets_dropped);
   // TODO(mattklein123): Handle no error when we limit the number of packets read.
-  // TODO(mattklein123): Increment stat on failure.
-  ASSERT(result->getErrorCode() == Api::IoError::IoErrorCode::Again);
+  if (result->getErrorCode() != Api::IoError::IoErrorCode::Again) {
+    // TODO(mattklein123): Upstream cluster RX error stat.
+  }
 }
 
 void UdpProxyFilter::ActiveSession::write(const Buffer::Instance& buffer) {
-  ENVOY_LOG(trace, "writing {} byte datagram: downstream={} local={} upstream={}", buffer.length(),
-            addresses_.peer_->asStringView(), addresses_.local_->asStringView(),
+  ENVOY_LOG(trace, "writing {} byte datagram upstream: downstream={} local={} upstream={}",
+            buffer.length(), addresses_.peer_->asStringView(), addresses_.local_->asStringView(),
             host_->address()->asStringView());
+  parent_.config_->stats().downstream_sess_rx_bytes_.add(buffer.length());
+  parent_.config_->stats().downstream_sess_rx_datagrams_.inc();
+
+  idle_timer_->enableTimer(parent_.config_->sessionTimeout());
```

Comment:
This is triggered once per incoming packet. Is it okay? 
Another way to check for idle timeout is that setting an alarm per loop and keeping a timestamp of last network activity. Each round, the alarm will check the last network activity timestamp against current time.

Reply:
I thought about implementing it the way that you propose, but this way is simpler and it's also how tcp proxy is implemented, for better or worse. I'm going to put in a TODO to revisit this for both UDP and TCP proxy.

Reply:
TODO sounds good. Just worth pointing it out that TCP proxy is not one read per packet. UDP doesn't have to either if we switch to recvmmsg. But udp callback is triggered once per packet.